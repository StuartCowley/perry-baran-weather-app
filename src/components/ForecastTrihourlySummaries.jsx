import React from "react";
import PropTypes from "prop-types";
import { v4 as uuid } from "uuid";
import "../styles/ForecastTrihourlySummaries.css";
import ForecastTrihourlySummary from "./ForecastTrihourlySummary";

function ForecastTriHourlySummaries({ forecasts }) {
  const formatForecasts = (trihourlyForecasts) => {
    const FORECAST_INTERVAL = 3;
    const FORECASTS_PER_DAY = 24 / FORECAST_INTERVAL;
    const forecastsArray = [];

    for (let i = 0; i < FORECASTS_PER_DAY; i += 1) {
      let time = `${i * FORECAST_INTERVAL}:00`;
      if (time.length === 4) time = `0${time}`;

      const trihourlyForecast = trihourlyForecasts.find(
        (forecast) => time === forecast.dateTimeTxt.slice(11, 16)
      );

      if (trihourlyForecast) {
        trihourlyForecast.time = time;
        forecastsArray.push(trihourlyForecast);
      } else {
        forecastsArray.push({
          time,
          dateTime: uuid(),
          weather: null,
          icon: null,
          temp: null,
        });
      }
    }

    return forecastsArray;
  };

  const trihourlyForecast = formatForecasts(forecasts);

  return (
    <div className="forecast-trihourly-summaries">
      {trihourlyForecast.map((forecast) => {
        const { time, dateTime, weather, icon, temp } = forecast;
        return (
          <ForecastTrihourlySummary
            time={time}
            weather={weather}
            icon={icon}
            temp={temp}
            key={dateTime}
          />
        );
      })}
    </div>
  );
}

ForecastTriHourlySummaries.propTypes = {
  forecasts: PropTypes.arrayOf(
    PropTypes.shape({
      dateTime: PropTypes.number,
      dateTimeTxt: PropTypes.string,
      icon: PropTypes.string,
      temp: PropTypes.number,
      maxTemp: PropTypes.number,
      minTemp: PropTypes.number,
      humidity: PropTypes.number,
      weather: PropTypes.string,
      windSpeed: PropTypes.number,
    })
  ).isRequired,
};

export default ForecastTriHourlySummaries;
