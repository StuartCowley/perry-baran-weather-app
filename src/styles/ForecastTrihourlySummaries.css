.forecast-trihourly-summaries {
  display: flex;
  justify-content: space-evenly;
}