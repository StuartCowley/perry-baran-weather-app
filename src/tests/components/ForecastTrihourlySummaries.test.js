import React from "react";
import { render } from "@testing-library/react";
import ForecastTrihourlySummaries from "../../components/ForecastTrihourlySummaries";

describe("ForecastTrihourlySummaries", () => {
  const forecasts = [
    {
      dateTime: 123,
      dateTimeTxt: "2022-11-03 15:00:00",
      icon: "200",
      temp: 12,
      maxTemp: 13,
      minTemp: 11,
      humidity: 44,
      weather: "really cold",
      windSpeed: 66,
    },
    {
      dateTime: 456,
      dateTimeTxt: "2022-11-03 18:00:00",
      icon: "500",
      temp: 22,
      maxTemp: 23,
      minTemp: 21,
      humidity: 45,
      weather: "less cold",
      windSpeed: 99,
    },
  ];
  let screen;

  beforeEach(() => {
    screen = render(<ForecastTrihourlySummaries forecasts={forecasts} />);
  });

  test("snapshot", () => {
    const { asFragment } = screen;

    expect(asFragment()).toMatchSnapshot();
  });

  test("ForecastTrihourlySummary is called correct number of times, for every 3 hours for 1 day", () => {
    const { getAllByTestId, getByText } = screen;
    const FORECAST_INTERVAL = 3;
    const FORECASTS_PER_DAY = 24 / FORECAST_INTERVAL;

    expect(getAllByTestId("forecast-trihourly-summary")).toHaveLength(
      FORECASTS_PER_DAY
    );

    for (let i = 0; i < FORECASTS_PER_DAY; i += 1) {
      let time = `${i * FORECAST_INTERVAL}:00`;
      if (time.length === 4) time = `0${time}`;
      expect(getByText(time)).toBeInTheDocument();
    }
  });
});
