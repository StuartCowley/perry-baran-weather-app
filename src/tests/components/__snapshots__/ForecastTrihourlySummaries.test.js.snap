// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastTrihourlySummaries snapshot 1`] = `
<DocumentFragment>
  <div
    class="forecast-trihourly-summaries"
  >
    <div
      class="forecast-trihourly-summary"
      data-testid="forecast-trihourly-summary"
    >
      <p>
        00:00
      </p>
      <p>
        No 
        <br />
        Data
      </p>
    </div>
    <div
      class="forecast-trihourly-summary"
      data-testid="forecast-trihourly-summary"
    >
      <p>
        03:00
      </p>
      <p>
        No 
        <br />
        Data
      </p>
    </div>
    <div
      class="forecast-trihourly-summary"
      data-testid="forecast-trihourly-summary"
    >
      <p>
        06:00
      </p>
      <p>
        No 
        <br />
        Data
      </p>
    </div>
    <div
      class="forecast-trihourly-summary"
      data-testid="forecast-trihourly-summary"
    >
      <p>
        09:00
      </p>
      <p>
        No 
        <br />
        Data
      </p>
    </div>
    <div
      class="forecast-trihourly-summary"
      data-testid="forecast-trihourly-summary"
    >
      <p>
        12:00
      </p>
      <p>
        No 
        <br />
        Data
      </p>
    </div>
    <div
      class="forecast-trihourly-summary"
      data-testid="forecast-trihourly-summary"
    >
      <p>
        15:00
      </p>
      <div
        class="forecast-trihourly-summary__icon"
        data-testid="forecast-icon"
      >
        <i
          class="wi wi-day-storm-showers"
          name="owm"
        />
      </div>
      <p>
        really 
        <br />
        cold
      </p>
      <p>
        12.0°C
      </p>
    </div>
    <div
      class="forecast-trihourly-summary"
      data-testid="forecast-trihourly-summary"
    >
      <p>
        18:00
      </p>
      <div
        class="forecast-trihourly-summary__icon"
        data-testid="forecast-icon"
      >
        <i
          class="wi wi-day-rain"
          name="owm"
        />
      </div>
      <p>
        less 
        <br />
        cold
      </p>
      <p>
        22.0°C
      </p>
    </div>
    <div
      class="forecast-trihourly-summary"
      data-testid="forecast-trihourly-summary"
    >
      <p>
        21:00
      </p>
      <p>
        No 
        <br />
        Data
      </p>
    </div>
  </div>
</DocumentFragment>
`;
